apps/aecore/src/aec_eoe_vote.erl

```diff
@@ -0,0 +1,512 @@
+-export([start_link/2, negotiate/8, get_finalize_transaction/1]).
+
+%% Export gen_statem callbacks
+-export([init/1, callback_mode/0, terminate/3, code_change/4]).
+
+%% FSM states (as per gen_statem callback callback_mode/0)
+-export([await_eoe/3, proposal/3, vote/3, finalize/3, complete/3]).
+
+
+%% ==================================================================
+%% Records and Types
+
+-record(data, {
+                epoch                      :: non_neg_integer() | undefined,
+                height                     :: non_neg_integer() | undefined,
+                fork_hash                  :: binary() | undefined,
+                seed                       :: binary() | undefined,
+                length                     :: non_neg_integer() | undefined,
+                length_delta               :: non_neg_integer() | undefined,
+                leader                     :: binary() | undefined,
+                validators=[]              :: [{binary(), non_neg_integer()}],
+                remaining_validators=#{}   :: #{binary() => non_neg_integer()},
+                stakers=#{}                :: #{binary() => binary()},
+                majority=0                 :: non_neg_integer(),
+                proposal                   :: #{binary() => any()} | undefined,
+                block_time                 :: non_neg_integer(),
+                result                     :: {ok, binary()} | {error, no_consensus} | undefined,
+                from                       :: pid() | undefined,
+                votes=#{}                  :: #{binary() => #{binary() => any()}}
+            }).
+
+
+-define(PROPOSAL_TYPE, 1).
+-define(VOTE_TYPE, 2).
+-define(COMMIT_TYPE, 3).
+
+-define(HASH_FLD, <<"block_hash">>).
+-define(HEIGHT_FLD, <<"block_height">>).
+-define(EPOCH_DELTA_FLD, <<"epoch_length_delta">>).
+-define(SIGNATURE_FLD, <<"signature">>).
+
+%% API to start the state machine
+-spec start_link(#{binary() => binary()}, non_neg_integer())  -> {ok, pid()} | {error, atom()}.
+start_link(Stakers, BlockTime) ->
+    gen_statem:start_link({local, ?MODULE}, ?MODULE, [Stakers, BlockTime], []).
+
+%% Negotiate a fork, called with preferred fork and epoch length delta
+-spec negotiate(non_neg_integer(), non_neg_integer(), binary(), aec_keys:pubkey(), [{binary(), non_neg_integer()}], binary(), non_neg_integer(), non_neg_integer()) -> ok.
+negotiate(Epoch, Height, Hash, Leader, Validators, Seed, LengthDelta, CurrentLength) ->
+    gen_statem:cast(?MODULE, {negotiate, Epoch, Height, Hash, Leader, Validators, Seed, LengthDelta, CurrentLength}).
+
+-spec get_finalize_transaction(aec_trees:trees()) -> {ok, aetx_sign:signed_tx()} | {error, not_ready} | {error, term()}.
+get_finalize_transaction(Trees) ->
+    gen_statem:call(?MODULE, {get_finalize_transaction, Trees}).
+
+%%% gen_statem callbacks
+
+%% Initialization: Start in the await_eoe state
+init([Stakers, BlockTime]) ->
+    aec_events:subscribe(tx_received),
```

Comment:
Yes, I think that would be more efficient. I have checked with other types of transactions. I noticed if the leader had a spend transaction in the last micro block the creation of the finalisation transaction failed because the nonce was incorrect. Now that I use the trees from aec_block_micro_candidate this is no longer an issue.